### PR TITLE
correct Classification Module

### PR DESCRIPTION
**Pull request by petroslk**
When pen_markings is used, the model used generates an output feature array of length H x W x 3. The coverslip_edge model generates an array of length H x W x 2. This output asymetry leads to an error when the "ClassificationModule.byExample:pen_markings" element of the pipeline is used in the config file because the reshape function is hardcoded to take an array of H x W x 2.

This if statement takes into account the difference in shape based on the name of the active module.

### CODE REVIEW

**Comment by nanli-emory** [comment]
Hi @petroslk & @choosehappy,

This issue should be related to #250. Please correct me if I misunderstood anything.
The output feature array is not always H x W x n_class. It depends on the # of classes. After I read the **RandomForestClassifier.predict_proba** document:
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.predict_proba
I realize that the n_class should be dynamic numbers. So the reshape function should be called by using the **n_classes** in **cal.shape(n_samples, n_classes)**
I think the code should be changed to:
```python
    cal = clf.predict_proba(feats.reshape(-1, feats.shape[2]))
    # get the classes #
    cal_classes = cal.shape[1]
    # reshape by predict classes
    cal = cal.reshape(img.shape[0], img.shape[1], cal_classes)
```

This is what I thought but I'm not 100% sure.

**Comment by choosehappy** [comment]
i'm confused by this... i think the error is somewhere else upstream that could be corrected so this if statement isn't needed

why is pen markings returning a  H x W x 3 and not an H x W x 2? shouldn't we *always* be doing binary classification?

**Comment by nanli-emory** [comment]
Hi 

> Hi @petroslk & @choosehappy,
> 
> This issue should be related to #250. Please correct me if I misunderstood anything. The output feature array is not always H x W x n_class. It depends on the # of classes. After I read the **RandomForestClassifier.predict_proba** document: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.predict_proba I realize that the n_class should be dynamic numbers. So the reshape function should be called by using the **n_classes** in **cal.shape(n_samples, n_classes)** I think the code should be changed to:
> 
> ```python
>     cal = clf.predict_proba(feats.reshape(-1, feats.shape[2]))
>     # get the classes #
>     cal_classes = cal.shape[1]
>     # reshape by predict classes
>     cal = cal.reshape(img.shape[0], img.shape[1], cal_classes)
> ```
> 
> This is what I thought but I'm not 100% sure.

Hi @jacksonjacobs1 , could you review the comment? The classes should be based on the type of pens. What do you think?

**Comment by choosehappy** [comment]
> Hi
> 
> > Hi @petroslk & @choosehappy,
> > This issue should be related to #250. Please correct me if I misunderstood anything. The output feature array is not always H x W x n_class. It depends on the # of classes. After I read the **RandomForestClassifier.predict_proba** document: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.predict_proba I realize that the n_class should be dynamic numbers. So the reshape function should be called by using the **n_classes** in **cal.shape(n_samples, n_classes)** I think the code should be changed to:
> > ```python
> >     cal = clf.predict_proba(feats.reshape(-1, feats.shape[2]))
> >     # get the classes #
> >     cal_classes = cal.shape[1]
> >     # reshape by predict classes
> >     cal = cal.reshape(img.shape[0], img.shape[1], cal_classes)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is what I thought but I'm not 100% sure.
> 
> Hi @jacksonjacobs1 , could you review the comment? The classes should be based on the type of pens. What do you think?

no, its a binary classification problem, either a pixel is predicted to be a pen marking, or it is not a pen marking, so the maximum dimensionality should be 2, as shown here:

https://github.com/choosehappy/HistoQC/blob/d5058ec96b3c9dcee8924853ca6c732449c9fbad/histoqc/ClassificationModule.py#L174

if the output dimension is >2, then something is wrong somewhere

does that make sense?

**Comment by jacksonjacobs1** [comment]
It looks like problem is the RandomForestClassifier has 3 classes instead of 2:
```python
>>> clf.classes_
array([  0.,   1., 255.])
```

I haven't checked, but I suspect that one or more mask loaded as uint8 instead of true binary. 

Well, I kind of checked:
```python
>>> model_labels.max()
255.
```

Probably has something to do with this line: https://github.com/choosehappy/HistoQC/blob/d5058ec96b3c9dcee8924853ca6c732449c9fbad/histoqc/ClassificationModule.py#L182C59-L182C59

@nanli-emory can you look into this?

**Comment by nanli-emory** [comment]
> It looks like problem is the RandomForestClassifier has 3 classes instead of 2:
> 
> ```python
> >>> clf.classes_
> array([  0.,   1., 255.])
> ```
> 
> I haven't checked, but I suspect that one or more mask loaded as uint8 instead of true binary.
> 
> Well, I kind of checked:
> 
> ```python
> >>> model_labels.max()
> 255.
> ```
> 
> Probably has something to do with this line: https://github.com/choosehappy/HistoQC/blob/d5058ec96b3c9dcee8924853ca6c732449c9fbad/histoqc/ClassificationModule.py#L182C59-L182C59
> 
> @nanli-emory can you look into this?

I will look into this. Will let you know if I fix the issue.

**Comment by nanli-emory** [comment]
I create a new PR #269 for fixing the bugs

**Comment by nanli-emory** [comment]
fixed this PR and bug in PR @269
